### PR TITLE
Mlton support

### PR DESCRIPTION
**Pull request by ghost**
The escape sequence stuff is kind of ugly but it seems to work and was the easiest thing to do for now. Mlton doesn't support the vector literals from sml/nj either apparently.

### CODE REVIEW

**Comment by ghost** [comment]
By the way, I'd be happy to go through and clean up the escape codes by putting them in a structure or something.

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 .cm
 .smackage
+bin
```
Comment (jonsterling) [inline .gitignore:3]: there's actually an important shell script in `bin` right now that is called by the `makefile`. Perhaps we should move it...

Comment (ghost) [inline .gitignore:3]: Ah, yeah that's fine. I just wanted to ignore the heapimage thing that gets generated.

**Comment by jonsterling** [comment]
Yes, I think it would be good to put the escape codes in a structure if we're going to replace the unicode with them...

My other slight concern is with regard to the lack of vector literals... In the refiner (see `ctt.fun`) there are hundreds of cases where things would have to get switched over, and it would likely become extremely noisy. Likewise, pattern matching becomes very problematic.

I think that one option would be to change the ABT implementation to use lists rather than vectors, though this is perhaps not ideal. But as much as I would like the implementation to be "universal", I don't want to introduce things that make it difficult or cumbersome to develop (or read). What do you think?

**Comment by ghost** [comment]
Yeah, I'm not sure what to do about that. I started trying to fix up `extract.fun` to compile and ran into a number of issues with the pattern matching already. But switching to lists just for MLton doesn't seem worth it to me. My main interest in having MLton support was for a slightly nicer story with regard to producing standlone binaries and also the fact that MLton has some LLVM support which could be kind of interesting for compiling realizers at some point. Neither of those are pressing issues of course, so if it turns out to be too difficult to make it work I'm okay with putting it aside.

Maybe I'll hold off on trying to make further changes along these lines unless you have some ideas about what to do with the vectors. I don't have enough SML experience to know what the alternatives are at this point.

**Comment by jonsterling** [comment]
Yeah... Now that JonPRL is no longer primarily meant to be used as a library, looking into compiling with MLTon does begin to make a lot of sense...

I think that at this point, my perspective would just be that there is a lot of stuff to be done before that side of things becomes the most pressing issue, though there is indeed a lot of interesting stuff to be done in that area. I guess I would say, if we can get MLTon compatibility without a really big amount of pain, we should do it—but I personally want to avoid getting mired in infrastructure hacking when I have not yet finished building a usable proof assistant.

**Comment by ghost** [comment]
Alright, I'm going to close this for now and maybe maintain it in a separate branch unless we figure out a good way to work around these issues.

**Pull request closed without merging**